paddle/go/pserver/client.go

```diff
@@ -1,12 +1,82 @@
 package pserver
 
+import (
+	"hash/fnv"
+	"log"
+	"sort"
+	"time"
+
+	"github.com/PaddlePaddle/Paddle/paddle/go/pserver/internal/connection"
+)
+
+// TODO(helin): add RPC call retry logic
+
+// Selector selects if the client should initialize parameter servers.
+type Selector interface {
+	Select() bool
+}
+
+// Server is the identification of a parameter Server.
+type Server struct {
+	Index int
+	Addr  string
+}
+
+// Lister lists currently available parameter servers.
+type Lister interface {
+	List() []Server
+}
+
 // Client is the client to parameter servers.
 type Client struct {
+	sel      Selector
+	pservers []*connection.Conn
 }
 
 // NewClient creates a new client.
-func NewClient(addr string) *Client {
-	return &Client{}
+func NewClient(l Lister, pserverNum int, sel Selector) *Client {
+	c := &Client{sel: sel}
+	c.pservers = make([]*connection.Conn, pserverNum)
+	for i := 0; i < pserverNum; i++ {
+		c.pservers[i] = connection.New()
+	}
+	go c.monitorPservers(l, pserverNum)
+	return c
+}
+
+// monitorPservers monitors pserver addresses, and updates connection
+// when the address changes.
+func (c *Client) monitorPservers(l Lister, pserverNum int) {
+	knownServers := make([]Server, pserverNum)
+	ticker := time.NewTicker(10 * time.Second)
+	monitor := func() {
+		curServers := make([]Server, pserverNum)
+		list := l.List()
```

Comment:
It's ok if number of pservers still did not reach the desired number. We will only connect to pservers that are alive. For pservers not alive yet, we will not connect to them, and all calls to them will be blocked until connected. (See: https://github.com/PaddlePaddle/Paddle/pull/2243/files#diff-b1a8f38187aa603e0efc16791f282d87R61)